tt_eager/tt_dnn/op_library/topk/kernels/topk.cpp

```diff
@@ -0,0 +1,171 @@
+    constexpr uint32_t input_transposed_cb_index = get_compile_time_arg_val(2);
+    constexpr uint32_t index_transposed_cb_index = get_compile_time_arg_val(3);
+    constexpr uint32_t values_cb_index = get_compile_time_arg_val(4);
+    constexpr uint32_t output_ind_cb_index = get_compile_time_arg_val(5);
+    constexpr uint32_t Ht = get_compile_time_arg_val(6);
+    constexpr uint32_t Wt = get_compile_time_arg_val(7);
+    constexpr uint32_t K = get_compile_time_arg_val(8);
+    constexpr uint32_t logk = get_compile_time_arg_val(9);
+    constexpr uint32_t logWt = get_compile_time_arg_val(10);
+
+    // init pack, compute and unpack
+    transpose_wh_init(input_cb_index, input_transposed_cb_index);
+    for(uint32_t ht = 0; ht < Ht; ++ht) {
+        bool ascending = false;
+        cb_reserve_back(input_transposed_cb_index, Wt);
+        cb_reserve_back(index_transposed_cb_index, Wt);
+
+        for (uint32_t wt = 0; wt < Wt; wt+=2) {
+            acquire_dst(tt::DstMode::Half);
+            // local sort into k groups
+            cb_wait_front(input_cb_index, 2);
+            cb_wait_front(index_cb_index, 2);
+
+            unpack_reconfig_data_format_srca(input_cb_index);
+            transpose_wh_init_short(input_cb_index);
+            transpose_wh_tile(input_cb_index, wt, 0);
+            transpose_wh_tile(input_cb_index, wt+1, 1);
+
+            unpack_reconfig_data_format_srca(index_cb_index);
+            transpose_wh_init_short(index_cb_index);
+            transpose_wh_tile(index_cb_index, wt, 2);
+            transpose_wh_tile(index_cb_index, wt+1, 3);
+
+            // llk_topk_sort -> inplace
+            ckernel::topk_tile_init();
+            ckernel::topk_local_sort(0, (int) ascending, logk - 1);
+
+            // pack value tiles into cb_intermed1
+            pack_reconfig_data_format(input_transposed_cb_index);
+            pack_tile(0, input_transposed_cb_index, wt);
+            pack_tile(1, input_transposed_cb_index, wt+1);
+
+            // pack index tiles into cb_intermed2
+            pack_reconfig_data_format(index_transposed_cb_index);
+            pack_tile(2, index_transposed_cb_index, wt);
+            pack_tile(3, index_transposed_cb_index, wt+1);
+
+            cb_pop_front(input_cb_index, 2);
+            cb_pop_front(index_cb_index, 2);
+
+            release_dst(tt::DstMode::Half);
+            ascending = !ascending;
+        }
+
+        cb_push_back(input_transposed_cb_index, Wt);
+        cb_push_back(index_transposed_cb_index, Wt);
+
+        cb_wait_front(input_transposed_cb_index, Wt);
+        cb_wait_front(index_transposed_cb_index, Wt);
+
+        cb_reserve_back(input_transposed_cb_index, Wt);
+        cb_reserve_back(index_transposed_cb_index, Wt);
+
+
+        // iterative merge and rebuild on pairs of tiles
+        constexpr uint32_t num_iterations = logWt;
+        copy_tile_to_dst_init_short_with_dt(index_cb_index, input_transposed_cb_index);
```

Comment:
Why do we re-init here? Aren't we coing to re-init immediately after in the loop?

Reply:
Two different datatypes between the CBs so I reinit

Reply:
Not sure I follow, line 87 does a copy re-init, but isn't this immediately followed by the re-init on line 96? So why would line 87 be needed? Or does it not always enter the following loops?
![image](https://github.com/tenstorrent/tt-metal/assets/109362939/a5ba5314-a226-46f5-ae08-af185e4c22e0)

Reply:
Oh, thought you were referring to the reinit in the loop. Yeah that first reinit is useless, good catch. Tested and removed.